https://github.com/sokrypton/ColabFold

 ColabFold accelerates protein structure and complex predictions by combining structure prediction models like AlphaFold2 or RoseTTAFold with a highly optimized, fast homology search tool called MMseqs2 (Many-against-Many sequence searching).

* The Fast Search Tool: Rather than using the slow default alignment tools from AlphaFold2 (like JackHMMER or HHblits), ColabFold substitutes them with MMseqs2. This homology search handles Multiple Sequence Alignment (MSA) generation up to 40–60 times faster.
 
*  Monomers and Complexes: It predicts 3D structures for both single-chain monomers and multi-subunit hetero/homo-complexes (multimers).
 
* Local and Cloud Execution: It can be run for free via Google Colaboratory or installed as a command-line tool (colabfold_batch or colabfold_search) to run locally on your own GPU resources

In [1]:
%pip install colabfold

Note: you may need to restart the kernel to use updated packages.


In [2]:
from colabfold.colabfold import run_mmseqs2

sequence = "MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQANL"
result = run_mmseqs2(
    [sequence],
    prefix="query",
    use_env=True,
    use_filter=True
)

print(type(result))
print(result)

No user agent specified. Please set a user agent (e.g., 'toolname/version contact@email') to help us debug in case of problems. This warning will become an error in the future.


<class 'list'>
['>101\nMKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQANL\n>UniRef100_A0A2X4ZW89\t62\t0.970\t1.421E-09\t0\t33\t36\t0\t33\t218\nMKTAYIAKQRQISFVKSHFSRQLEEKLGLIEVQA--\n>UniRef100_A0A4U9TP15\t61\t0.764\t1.955E-09\t0\t33\t36\t0\t33\t46\nMKKQFIQKQQQISFVKSFFSRQLEQQLGLIEVQA--\n>UniRef100_A0A2X1QBJ4\t61\t1.00\t1.955E-09\t0\t33\t36\t0\t33\t51\nMKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--\n>UniRef100_A0A379X549\t61\t1.00\t1.955E-09\t0\t33\t36\t0\t33\t83\nMKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--\n>UniRef100_A0A2X3KQB2\t61\t1.00\t1.955E-09\t0\t33\t36\t0\t33\t185\nMKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--\n>UniRef100_A0A378MTS9\t59\t0.515\t9.623E-09\t0\t32\t36\t0\t32\t86\nMKKSFILQQQEISFVKNTFTQYLIDKLGIIEVQ---\n>UniRef100_A0A377PRM1\t59\t0.764\t9.623E-09\t0\t33\t36\t0\t33\t89\nMKKQFIQKQQQISFVKSSFSRQLEKQLGLIEVQA--\n>UniRef100_A0A654U887\t59\t0.454\t1.821E-08\t0\t32\t36\t0\t32\t89\nMKKSFIHQQEEISFVKNTFTQYLKDKLEVVEVQ---\n>UniRef100_A0A7H4MAC1\t56\t0.967\t1.234E-07\t0\t30\t36\t0\t30\t36\nMKTAYIAKQRQISFVKSHFSRQLEEKLGLIE--

run_mmseqs2() sent 37-residue input (MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQANL) to the MMseqs2 web server (api.colabfold.com), which searched it against two databases (since use_env=True):

- UniRef — curated/clustered protein sequences (hits labeled UniRef100_...)
- Environmental/metagenomic database — things like MGnify protein sequences (hits labeled MGYP...), plus assembled metagenome scaffolds

run_mmseqs2() does two different things:

[1.] Sequence homology search (for building the MSA) → always searches uniref30 + colabfold_envdb (if use_env=True)

[2.] Structural template search (for finding known 3D structures to use as folding templates) 

→ searches pdb100 (the searchable/clustered version of the Protein Data Bank), but only if you explicitly set use_templates=True

In [3]:
msa = result[0]

print(msa)

>101
MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQANL
>UniRef100_A0A2X4ZW89	62	0.970	1.421E-09	0	33	36	0	33	218
MKTAYIAKQRQISFVKSHFSRQLEEKLGLIEVQA--
>UniRef100_A0A4U9TP15	61	0.764	1.955E-09	0	33	36	0	33	46
MKKQFIQKQQQISFVKSFFSRQLEQQLGLIEVQA--
>UniRef100_A0A2X1QBJ4	61	1.00	1.955E-09	0	33	36	0	33	51
MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--
>UniRef100_A0A379X549	61	1.00	1.955E-09	0	33	36	0	33	83
MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--
>UniRef100_A0A2X3KQB2	61	1.00	1.955E-09	0	33	36	0	33	185
MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQA--
>UniRef100_A0A378MTS9	59	0.515	9.623E-09	0	32	36	0	32	86
MKKSFILQQQEISFVKNTFTQYLIDKLGIIEVQ---
>UniRef100_A0A377PRM1	59	0.764	9.623E-09	0	33	36	0	33	89
MKKQFIQKQQQISFVKSSFSRQLEKQLGLIEVQA--
>UniRef100_A0A654U887	59	0.454	1.821E-08	0	32	36	0	32	89
MKKSFIHQQEEISFVKNTFTQYLKDKLEVVEVQ---
>UniRef100_A0A7H4MAC1	56	0.967	1.234E-07	0	30	36	0	30	36
MKTAYIAKQRQISFVKSHFSRQLEEKLGLIE-----
>UniRef100_A0A376LE91	53	1.00	1.153E-06	0	29	36	0	29	37
MKTAYIAKQRQISFVKSHFSRQLEERLGLI------
>UniRef100_A0A2X3EC4

>UniRef100_A0A2X4ZW89	62	0.970	1.421E-09	0	33	36	0	33	218
MKTAYIAKQRQISFVKSHFSRQLEEKLGLIEVQA--

This is MMseqs2's tabular output format (similar to BLAST's format 6), with 10 tab-separated fields:
| # | Value | Field Name | Meaning |
|---|-------|------------|---------|
| 1 | `UniRef100_A0A2X4ZW89` | Target ID | The database accession of the matched sequence |
| 2 | `62` | Bit Score | Alignment quality score, normalized for statistical comparison across searches. Higher = stronger match |
| 3 | `0.970` | Sequence Identity | Fraction of aligned residues that are identical between your query and this hit → 97% identical |
| 4 | `1.421E-09` | E-value | Expected number of hits this good you'd see by pure chance in a database this size. Lower = more significant |
| 5 | `0` | Query Start | Position in your query sequence where the alignment begins (0-indexed) |
| 6 | `33` | Query End | Position in your query sequence where the alignment ends |
| 7 | `36` | Query Length | Total length of your query sequence (37 aa, but 0-indexed so max index is 36) |
| 8 | `0` | Target Start | Position in the matched database sequence where the alignment begins |
| 9 | `33` | Target End | Position in the matched database sequence where the alignment ends |
| 10 | `218` | Target Length | Total length of the full database sequence (this hit protein is 218 aa long, but only residues 0–33 aligned to your 37-aa query) |

In [4]:
with open("query.a3m", "w") as f:
    f.write(msa)

print("Saved:", "query.a3m")

Saved: query.a3m


In [ ]:
import numpy as np
import re
import sys

'''
1. Encode the input sequences to numberals based on the index identified.
2. Step2: Once you have encoded sequence, down-weight the near-duplicate sequences to avoid dominance of similar signals in co-variance matrix
3. Step3: Fetch Single and Pair frequencies..
4. Step4: Fetch direct couplings - DCA analysis
    4a. Build the (L*(Q-1)) x (L*(Q-1)) covariance matrix (dropping the gap state as reference), (LX(Q-1)) X (L(Q-1))
    4b. invert it (mean-field), and extract per-pair coupling strength via Frobenius norm -> raw DI-like score.
5. Average Product Correction (removes background phylogenetic noise) [relating to the evolutionary history and development of a group of organisms. It describes how different species or groups are related to each other through common ancestry]
     If two species split from a recent common ancestor, their entire proteins will look highly similar. 
     If position X and position Y happen to match in both species, it is not because they are physically 
     interacting—it is simply because they both inherited those positions from the same ancestor
'''

AA_ALPHABET = "-ACDEFGHIKLMNPQRSTVWY"  # '-' = gap, index 0
AA_TO_IDX = {aa: i for i, aa in enumerate(AA_ALPHABET)}
Q = len(AA_ALPHABET)  # 21 states (20 aa + gap)


def parse_a3m(a3m_text):
    """
    Parse an a3m alignment string into a list of aligned sequences,
    keeping only match-state columns (uppercase + '-'), stripping
    insert-state columns (lowercase letters and '.').
    Returns list of equal-length strings.
    """
    lines = a3m_text.strip().splitlines()
    seqs = []
    current = []
    for line in lines:
        if line.startswith(">"):
            if current:
                seqs.append("".join(current))
                current = []
        else:
            current.append(line.strip())
    if current:
        seqs.append("".join(current))

    # Strip lowercase (insertions) and '.' -- keep only match columns
    cleaned = []
    for s in seqs:
        s = re.sub(r"[a-z\.]", "", s)
        cleaned.append(s)

    # Sanity check: all sequences must now be the same length
    lengths = {len(s) for s in cleaned}
    if len(lengths) != 1:
        # Trim/pad to the query length (first sequence) as a fallback
        L = len(cleaned[0])
        cleaned = [s[:L].ljust(L, "-") for s in cleaned]

    return cleaned


def encode_msa(seqs):
    """Convert list of aligned sequences into an (N_seq x L) integer array."""
    L = len(seqs[0])
    N = len(seqs)
    msa = np.zeros((N, L), dtype=np.int8)
    for i, s in enumerate(seqs):
        for j, aa in enumerate(s):
            msa[i, j] = AA_TO_IDX.get(aa.upper(), 0)  # unknown -> gap
    return msa


def sequence_weights(msa, threshold=0.8):
    """
    Down-weight near-duplicate sequences (standard DCA reweighting),
    so a big cluster of near-identical hits doesn't dominate the
    covariation signal. threshold = fractional identity cutoff.
    """
    N, L = msa.shape
    weights = np.ones(N)
    # Pairwise identity in blocks to keep memory reasonable
    for i in range(N):
        sim = (msa == msa[i]).mean(axis=1)
        n_similar = np.sum(sim >= threshold)
        weights[i] = 1.0 / n_similar
    return weights


def pair_frequencies(msa, weights, pseudocount=0.5):
    """Compute single-site and pairwise amino-acid frequencies with pseudocounts."""
    N, L = msa.shape
    Meff = weights.sum() #smoothing factor

    f_single = np.zeros((L, Q)) # LX21 length
    for a in range(Q):
        f_single[:, a] = (weights[:, None] * (msa == a)).sum(axis=0) #This creates a Boolean array indicating where amino acid a occurs.
    f_single = f_single / Meff
    f_single = (1 - pseudocount) * f_single + pseudocount / Q

    #f_ij(a,b) = (1/N) * Σ_n  x_i^(n)(a) · x_j^(n)(b)
    f_pair = np.zeros((L, L, Q, Q)) #It's LXL only nut each LXL is QXQ storing - Frequency of amino acid a at position i AND amino acid b at position j.
    for a in range(Q):
        for b in range(Q):
            match_a = (msa == a).astype(float)
            match_b = (msa == b).astype(float)
            f_pair[:, :, a, b] = (weights[:, None, None] *
                                  match_a[:, :, None] *
                                  match_b[:, None, :]).sum(axis=0)
    f_pair /= Meff
    f_pair = (1 - pseudocount) * f_pair + pseudocount / (Q * Q)

    return f_single, f_pair, Meff


def mean_field_couplings(f_single, f_pair, L, reg=1e-3):
    """
    Build the (L*(Q-1)) x (L*(Q-1)) covariance matrix (dropping the gap
    state as reference), invert it (mean-field), and extract per-pair
    coupling strength via Frobenius norm -> raw DI-like score.
    """
    ###### Co-variance matrix is same size as f_pair/f_ij as  C_ij(a,b) = f_ij(a,b) − f_i(a) · f_j(b) - (L*(Q-1)) x (L*(Q-1))

    q = Q - 1  
    size = L * q
    C = np.zeros((size, size))

    def idx(i, a):
        return i * q + a

    for i in range(L):
        for a in range(1, Q):
            for j in range(L):
                for b in range(1, Q):
                    val = f_pair[i, j, a, b] - f_single[i, a] * f_single[j, b] #Co-variance matrix
                    if i == j and a == b:
                        val = f_single[i, a] * (1 - f_single[i, a])
                    C[idx(i, a - 1), idx(j, b - 1)] = val

    C += reg * np.eye(size)  # regularize for stable inversion
    C_inv = np.linalg.inv(C)
    e_ij = -C_inv  #couplings

    #The scores part is taking the large inverse covariance matrix and converting it into a simple L×L matrix of coupling strengths between positions.
    #The covariance matrix C tells you how the amino-acid states at different positions in the MSA vary together.
    scores = np.zeros((L, L))
    for i in range(L):
        for j in range(L):
            if i == j: #diagonal values
                continue
            block = e_ij[idx(i, 0):idx(i, 0) + q, idx(j, 0):idx(j, 0) + q] #It takes all the amino-acid-level couplings between two positions and summarizes them into one coupling-strength score.
            '''
                Extract their coupling block from e_ij.
                This block is (Q-1) × (Q-1).
            '''
            scores[i, j] = np.linalg.norm(block, ord="fro") #Frobenius norm

    return scores


def apc_correction(raw_scores):
    """Average Product Correction (removes background phylogenetic noise)."""
    L = raw_scores.shape[0]
    row_mean = raw_scores.sum(axis=1) / (L - 1)
    total_mean = raw_scores.sum() / (L * (L - 1))
    apc = np.outer(row_mean, row_mean) / total_mean
    corrected = raw_scores - apc
    np.fill_diagonal(corrected, 0)
    return corrected


def run_dca_from_a3m(a3m_text, max_seqs=2000, min_seqs=10):
    '''
        
        Step1: Is to encode the input sequence - (LXNseq) length, the index is defined above AA_ALPHABET, AA_TO_IDX.
        Step2: Once you have encoded sequence, down-weight the near-duplicate sequences to avoid dominance of similar signals in co-variance matrix
        Step3: Fetch Single and Pair frequencies..
        Step4: Fetch direct couplings - DCA analysis
    '''

    seqs = parse_a3m(a3m_text)
    if len(seqs) > max_seqs:
        seqs = seqs[:max_seqs]  # simple cap for speed; keep top hits (usually best-scoring first)
    if len(seqs) < min_seqs:
        raise ValueError(
            f"Only {len(seqs)} sequences after parsing -- too few for reliable "
            f"DCA. Coevolution signal needs a deep MSA (ideally 100s-1000s of "
            f"diverse sequences)."
        )

    msa = encode_msa(seqs) #STEP1 - ENCODE THE SEQUENCES
    weights = sequence_weights(msa)  #Sanity Step to Down-weight near-duplicate sequences (standard DCA reweighting) to avoid dominance of similar signals.
    f_single, f_pair, Meff = pair_frequencies(msa, weights)
    print(f"Parsed {len(seqs)} sequences, length {msa.shape[1]}, "
          f"effective sequence count (Meff) = {Meff:.1f}")

    raw_scores = mean_field_couplings(f_single, f_pair, msa.shape[1])
    di_matrix = apc_correction(raw_scores)
    return di_matrix, list(range(1, msa.shape[1] + 1))  # 1-indexed positions


def top_pairs(di_matrix, positions, top_n=20, min_seq_sep=4):
    """Return the top_n highest-scoring residue pairs, excluding pairs
    that are trivially close in sequence (min_seq_sep apart)."""
    L = di_matrix.shape[0]
    pairs = []
    for i in range(L):
        for j in range(i + 1, L):
            if j - i >= min_seq_sep:
                pairs.append((positions[i], positions[j], di_matrix[i, j]))
    pairs.sort(key=lambda x: -x[2])
    return pairs[:top_n]


def plot_contact_map(di_matrix, positions, top_n=20, out_path="contact_map.png"):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    fig, ax = plt.subplots(figsize=(6, 6))
    im = ax.imshow(di_matrix, cmap="viridis", origin="lower")
    ax.set_xlabel("Residue position")
    ax.set_ylabel("Residue position")
    ax.set_title("Predicted coevolution (DCA) contact map")
    fig.colorbar(im, ax=ax, label="DI score (APC-corrected)")
    fig.tight_layout()
    fig.savefig(out_path, dpi=150)
    print(f"Saved contact map to {out_path}")

    print(f"\nTop {top_n} coevolving residue pairs (sequence separation >= 4):")
    for i, j, score in top_pairs(di_matrix, positions, top_n=top_n):
        print(f"  {i:>4} -- {j:<4}  DI = {score:.3f}")


if __name__ == "__main__":
    with open("query.a3m") as f:
        a3m_text = f.read()

    di_matrix, positions = run_dca_from_a3m(a3m_text)
    plot_contact_map(di_matrix, positions, top_n=20, out_path="contact_map.png")

Parsed 112 sequences, length 36, effective sequence count (Meff) = 24.5
Saved contact map to contact_map.png

Top 20 coevolving residue pairs (sequence separation >= 4):
    18 -- 25    DI = 4.431
    24 -- 28    DI = 4.371
     3 -- 8     DI = 3.306
     9 -- 16    DI = 3.059
    29 -- 34    DI = 3.008
     4 -- 8     DI = 2.963
     1 -- 5     DI = 2.901
    17 -- 24    DI = 2.825
     4 -- 10    DI = 2.576
     2 -- 6     DI = 2.563
    27 -- 31    DI = 2.549
     3 -- 7     DI = 2.546
    16 -- 29    DI = 2.477
    17 -- 21    DI = 2.382
     6 -- 21    DI = 2.318
    28 -- 33    DI = 2.257
    10 -- 25    DI = 2.199
    23 -- 31    DI = 2.177
    29 -- 33    DI = 2.103
    14 -- 19    DI = 2.065


In [ ]:
"""
This is NOT a real protein structure prediction. It takes DCA coupling
scores (residue pairs likely to be spatially close, based on
coevolution) and embeds all residues into 3D using classical
multidimensional scaling (MDS), constrained loosely to look like a
connected polypeptide chain. Real geometry, bond angles, secondary
structure, and physical plausibility are NOT enforced.

Usage:
    from coevolution_dca import run_dca_from_a3m
    from dca_3d_embedding import build_distance_matrix, classical_mds, plot_3d_trace

    di_matrix, positions = run_dca_from_a3m(a3m_text)
    dist = build_distance_matrix(di_matrix)
    coords = classical_mds(dist, n_components=3)
    plot_3d_trace(coords, di_matrix, positions, top_n=15, out_path="rough_3d.png")
"""

import numpy as np


def build_distance_matrix(di_matrix, backbone_dist=3.8, max_dist=20.0,
                           backbone_window=3):
    """
    Convert a DI (coevolution) score matrix into a pseudo-distance matrix: strong coupling -> short distance, weak coupling -> long
    distance. Also softly enforces that residues close in sequence (within `backbone_window`) stay near a realistic backbone spacing,
    so the embedding looks like a connected chain rather than a scattered point cloud.
    """
    L = di_matrix.shape[0]
    di = di_matrix.copy()
    np.fill_diagonal(di, di.max())  # self-distance will be forced to 0 anyway

    di_min, di_max = di.min(), di.max()
    di_norm = (di - di_min) / (di_max - di_min + 1e-9)
    dist = max_dist * (1 - di_norm)

    for i in range(L):
        for j in range(L):
            sep = abs(i - j)
            if sep == 0:
                dist[i, j] = 0.0
            elif sep <= backbone_window:
                target = backbone_dist * sep
                dist[i, j] = 0.7 * target + 0.3 * dist[i, j]

    dist = (dist + dist.T) / 2
    return dist


def classical_mds(dist_matrix, n_components=3):
    """Classical (Torgerson) MDS: embed points in n_components dims
    such that pairwise Euclidean distances approximate dist_matrix."""
    n = dist_matrix.shape[0]
    D2 = dist_matrix ** 2
    J = np.eye(n) - np.ones((n, n)) / n
    B = -0.5 * J @ D2 @ J

    eigvals, eigvecs = np.linalg.eigh(B)
    order = np.argsort(eigvals)[::-1][:n_components]
    top_vals = np.clip(eigvals[order], 0, None)
    coords = eigvecs[:, order] * np.sqrt(top_vals)
    return coords


def plot_3d_trace(coords, di_matrix, positions, top_n=15, out_path="rough_3d.png"):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
    from mpl_toolkits.mplot3d import Axes3D  # noqa: F401

    fig = plt.figure(figsize=(7, 7))
    ax = fig.add_subplot(111, projection="3d")

    xs, ys, zs = coords[:, 0], coords[:, 1], coords[:, 2]

    # Backbone trace connecting residues in sequence order
    ax.plot(xs, ys, zs, color="gray", linewidth=1.5, alpha=0.7, zorder=1)

    # Color residues along the chain (N->C rainbow) so you can orient yourself
    sc = ax.scatter(xs, ys, zs, c=positions, cmap="viridis", s=40, zorder=2)
    fig.colorbar(sc, ax=ax, label="Residue position", shrink=0.6, pad=0.1)

    # Overlay top DCA-predicted contacts as dashed lines
    pairs = top_pairs(di_matrix, positions, top_n=top_n, min_seq_sep=4)
    for i, j, score in pairs:
        idx_i, idx_j = positions.index(i), positions.index(j)
        ax.plot(
            [xs[idx_i], xs[idx_j]],
            [ys[idx_i], ys[idx_j]],
            [zs[idx_i], zs[idx_j]],
            color="red", linestyle="--", linewidth=1, alpha=0.6, zorder=3,
        )

    ax.set_title("Rough DCA-based 3D embedding (illustrative only, NOT a real structure)")
    ax.set_xlabel("x (arb. units)")
    ax.set_ylabel("y (arb. units)")
    ax.set_zlabel("z (arb. units)")

    fig.tight_layout()
    fig.savefig(out_path, dpi=150)
    print(f"Saved rough 3D embedding to {out_path}")
    print(f"Overlaid top {len(pairs)} DCA-predicted contacts as red dashed lines")


if __name__ == "__main__":
    with open("query.a3m") as f:
        a3m_text = f.read()

    di_matrix, positions = run_dca_from_a3m(a3m_text)
    dist = build_distance_matrix(di_matrix)
    coords = classical_mds(dist, n_components=3)
    plot_3d_trace(coords, di_matrix, positions, top_n=5, out_path="output.png")

Parsed 112 sequences, length 36, effective sequence count (Meff) = 24.5
Saved rough 3D embedding to output.png
Overlaid top 5 DCA-predicted contacts as red dashed lines


In [8]:
from colabfold.colabfold import run_mmseqs2

sequence = "MKTAYIAKQRQISFVKSHFSRQLEERLGLIEVQANL"
result = run_mmseqs2([sequence], prefix="query", use_env=True, use_filter=True)
a3m_text = result[0]

di_matrix, positions = run_dca_from_a3m(a3m_text)
dist = build_distance_matrix(di_matrix)
coords = classical_mds(dist, n_components=3)
plot_3d_trace(coords, di_matrix, positions, top_n=15, out_path="rough_3d.png")

No user agent specified. Please set a user agent (e.g., 'toolname/version contact@email') to help us debug in case of problems. This warning will become an error in the future.


Parsed 112 sequences, length 36, effective sequence count (Meff) = 24.5
Saved rough 3D embedding to rough_3d.png
Overlaid top 15 DCA-predicted contacts as red dashed lines
